### DISCUSSION TITLE
Rework Point Calculation System

### DISCUSSION THREAD

**Discussion by vainamov** [Traewelling/traewelling#2008] [Ideas / Feature Requests]
**Is your feature request related to a problem? Please describe.**
Currently, the points awarded for any check-in, are the sum of the base points for the used vehicle type and 1 point for each commenced 10km. This leads to the odd situation, where a single check-in between stations that are less than 10km apart awards you less points than multiple.

When I travel 8 km from _Hauptbahnhof (U), Hannover_ to _Eichstraße (Bahnhof), Laatzen_ using the direct tram line 1, I receive 3 points (2 base points + 1 for 8km travelled). Let's say, because it's often less crowded and the trains used are newer, I take the line 8 until Bothmerstraße, Hannover first and than from there the line 1 for the last stations. Suddenly I receive 6 points (2 base points + 1 for 6km and 2 base points + 1 for 2km) although neither my travel time nor the distance increased, even the mean of transport stayed the same.

There lies another issue in there, as most inner city trips are less than 10km. It makes it near impossible to get more than the base points + 1 but also determines a 500m bus trip is worth the same as 20 minutes in a tram across 9km.

**Describe the solution you'd like**
I'm not sure that there even is _the_ solution to this, mainly because I imagine there being a lot of different opinions about this. But nonetheless I've come up with something, that I think might help mitigate the effect of this. (I excluded Taxis from this by choice, because I've never seen a check-in to a taxi and honestly wouldn't even consider taking taxis something worth awarding points for in the first place.)

Anyways, here's my plan:

- Strictly combine the transport types into categories. **Platinum** - long-distance, **Gold** - regional, **Silver** - remaining suburban and urban transport systems. (This would leave e.g. a **Bronze** option, for non-public transport types like some discussed here: https://github.com/Traewelling/traewelling/issues/1518#issuecomment-1583101319, but the whole naming is purely for demonstration)

- Award a point for each commenced kilometer to the respective category. Travelled 160km with an ICE? 160 Platinum-Points. 9km with a tram? 9 Silver-Points. 500m in bus? 1 Silver-Point.

- To calculate a total number of points the points from each category could be scaled and then added together. Every Platinum-, every 2 Gold-, every 3 Silver- and maybe every 5 Bronze-Points are worth 1 total point. Of course, the values could and probably should be tweaked to achieve a better balance but this could be done at any time, because it wouldn't affect how the (base) points are calculated and stored in the DB.

- To accommodate the current "punishments" for early/late check-ins it would either be possible to simply award the points to a lower category or apply the same percentage rules like now to the base points per km.

I think a system like this could help to achieve a fair distribution of points while maintaining the ratios between "lower and higher-class" transportation. It leads to higher points than currently but for one, higher values look nicer and are (to me at least) more motivating, but also the points still expire after 7 days so it should stay within limits.

**Additional context**
This change would also allow seperate leaderboards for the different categories. Furthermore it could also help with #1634.

**Comment by MrKrisKrisu** [comment]
I'm now increasingly thinking about revising the points system, as the points distribution is very unfair.

A simple change might be to decouple the calculation from the kilometers and focus more on the duration. So that the points are not rewarded for just the journey, but also for the correct use of Träwelling, you could perhaps consider giving more points the closer you check in to the departure time.

e.g. (fictional example to illustrate, no realistic values)
-> 60 minutes before departure time: no bonus points
-> 30 minutes before departure time: Increase the points factor to 1.5
-> 15 minutes before departure time: factor of 1.75
-> at departure time: factor of 2

I'm looking forward to further ideas.

**Reply by vainamov** [reply]
In general I think duration is an awesome metric. Of course I can get much further within a given timeframe on a high-speed train than on a slow bus, but the actual time spent using public transportation is the same, so why shouldn't the points be?

However we might need to consider some technicalities for this to work reliably and fair.
- Duration is prone to fluctuate. Delays or made up time can change the actual travel time and it's not unusual, that this happens during an active check-in of a user. My current solution would be to not award points until long after the destination was reached (or show them in an actively withhold state until then, to inform the user that the value might still change). Obviously there are other solutions e.g. considering only the planned duration or only the estimated duration at the time of check-in but those might be less fair.
- Duration can easily be changed by the user. Could and would we want to prevent exaggerated manual delays?
- Regarding the increasing factors closer to the departure time:
  - Just considering the actual departure wouldn't be fair, because if I check-in to a planned connection that gets delayed, I shouldn't be punished because the actual departure changes.
  - Just considering the planned departure wouldn't be fair, beacuse if I check-in to a delayed connection (that happens to be a connection I can take because of that delay), I shouldn't be punished either.

  So we should find a solution to account for both departure times and value them equally.

**Reply by bendix-dev** [reply]
I'm definitely in favor of using duration instead of distance. 

I'm not sure about penalising checking in early though, I don't see how it's a bad thing to check in early. I would prefer giving full points until the journey starts and then only penalise for checking in late. Also consider journeys where the vehicle has quite some standing time at the start point.

**Comment by pascalwittler** [comment]
A scoring system based on 100% time or 100% distance seems useless to me since kilometers and hours spent on trips are actually the two other numbers you can already see in the profile header next to your points. Personally, I do not need another scoring system at all, as you could simply display "total kilometers", "kilometers this week", "time spent at all" and "time spent this week" in a profile.

The idea of rewarding time spent on a trip seems quite nice at first glance but the only technically implementable metric would be time spent on a train, so many people would get less points than they would deserve if they have to wait a long time at a station or if there are delays or cancellations.

A mix of multiple metrics/factors combined relatively (instead of an absolute number of points on top) as described by @vainamov sounds great to me and would solve a specific, obvious problem with the current calculation system.

The aim of making the points system "fair" will remain an illusion at all as long as check-ins by train staff, business trips and private trips are rewarded with the same amount of points and as long as people are not told what the goal of the game is and how many points you get for what or when and why you lose them again.

**Comment by HerrLevin** [comment]
I've been thinking a bit about the possibilities of a new point calculation system. I have to agree, that the most "fair" approach would be to make the points based both on distance and duration. But I've tried to simplify this approach a bit.

In this scenario we would have three broad categories:
 - (fast) long distance transport: ICE, IC, FLX, EC, etc.
 - regional transport: IRE, RE, RB, etc.
 - "slow" transport: S-Bahn, Tram, U-Bahn, Bus

The main idea would be, that all categories are calculated based on time but each with their own formula. The long distance would start off at a higher baseline but both lower categories would slowly overtake the higher one due to a quadratic function.

I've graphed it out at https://www.desmos.com/calculator/kjbodrkvx3 -- it's far from perfect, so any feedback would be welcome!


![telegram-cloud-photo-size-2-5310251742224573272-y](https://github.com/user-attachments/assets/24e32655-4dcd-4249-9cc6-fa0fd42141bd)
